### PR TITLE
gh-85287: Convert UnicodeError to UnicodeEncodeError| UnicodeDecodeError 

### PR DESCRIPTION
**Pull request by srinivasreddy**
https://bugs.python.org/issue41115

* Issue: gh-85287

### CODE REVIEW

File: Lib/encodings/punycode.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-""" Codec for the Punicode encoding, as specified in RFC 3492
+""" Codec for the Punycode encoding, as specified in RFC 3492
```
Comment (eamanu) [inline Lib/encodings/punycode.py:1]: good catch

**Review by eamanu — APPROVED** [review]
LGTM

**Comment by doerwalter** [comment]
The patch get the offsets in the exception object wrong for all non-trivial cases.

For example the following case:

```python
>>> s = 'foo.' + 60*'\xff'
>>> s.encode('idna')
```

gives the stack trace:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/walter/checkouts/cpython-doerwalter/Lib/encodings/idna.py", line 187, in encode
    result.extend(ToASCII(label))
  File "/Users/walter/checkouts/cpython-doerwalter/Lib/encodings/idna.py", line 105, in ToASCII
    raise UnicodeEncodeError("punycode", label.decode("punycode"), 0,
UnicodeEncodeError: 'punycode' codec can't encode characters in position 0-62: label too long
```

However the bad characters are not `s[0:63]` but `s[4:64]`.

Fixing this would require to track the correct offset across multiple function calls. I'm not sure whether that added complexity is justified. `idna` enforces strict encoding anyway.

**Comment by arhadthedev** [comment]
A review by @doerwalter hasn't been being addressed for three years and the OP blocked their branch from external pushes thus preventing us from applying all necessary fixes by ourselves.

**Pull request closed without merging**